src/lightmap.cpp

```diff
@@ -882,43 +882,163 @@ static constexpr quadrant quadrant_from_x_y( int x, int y )
            ( ( y > 0 ) ? quadrant::NE : quadrant::SE );
 }
 
-// Add defaults for when method is invoked for the first time.
-template<int xx, int xy, int xz, int yx, int yy, int yz, int zz, typename T,
-         T( *calc )( const T &, const T &, const int & ),
-         bool( *check )( const T &, const T & ),
-         T( *accumulate )( const T &, const T &, const int & )>
-void cast_zlight_segment(
-    const array_of_grids_of<T> &output_caches,
-    const array_of_grids_of<const T> &input_arrays,
-    const array_of_grids_of<const bool> &floor_caches,
-    const array_of_grids_of <const diagonal_blocks> &blocked_caches,
-    const tripoint &offset, int offset_distance,
-    T numerator = 1.0f, int row = 1,
-    float start_major = 0.0f, float end_major = 1.0f,
-    float start_minor = 0.0f, float end_minor = 1.0f,
-    T cumulative_transparency = LIGHT_TRANSPARENCY_OPEN_AIR,
-    int x_skip = -1, int z_skip = -1 );
-
-template<int xx, int xy, int xz, int yx, int yy, int yz, int zz, typename T,
+struct slope {
+    slope( int_least8_t rise, int_least8_t run ) {
+        // Ensure run is always positive for the inequality operators
+        this->run = std::abs( run );
+        if( run < 0 ) {
+            this->rise = -rise;
+        } else {
+            this->rise = rise;
+        }
+    }
+    int_least8_t rise;
+    int_least8_t run;
+};
+
+static bool operator<( const slope &lhs, const slope &rhs )
+{
+    // a/b < c/d <=> a*d < c*b if b and d have the same sign.
+    return lhs.rise * rhs.run < rhs.rise * lhs.run;
+}
+static bool operator>( const slope &lhs, const slope &rhs )
+{
+    return rhs < lhs;
+}
+static bool operator==( const slope &lhs, const slope &rhs )
+{
+    // a/b == c/d <=> a*d == c*b
+    return lhs.rise * rhs.run == rhs.rise * lhs.run;
+}
+
+template<typename T>
+struct span {
+    span( const slope &s_major, const slope &e_major,
+          const slope &s_minor, const slope &e_minor,
+          const T &value, bool skip_first_row = false, bool skip_first_column = false ) :
+        start_major( s_major ), end_major( e_major ), start_minor( s_minor ), end_minor( e_minor ),
+        cumulative_value( value ), skip_first_row( skip_first_row ),
+        skip_first_column( skip_first_column ) {}
+    slope start_major;
+    slope end_major;
+    slope start_minor;
+    slope end_minor;
+    T cumulative_value;
+    bool skip_first_row;
+    bool skip_first_column;
+};
```

Comment:
```suggestion
template<typename T>
struct span {
    slope start_major;
    slope end_major;
    slope start_minor;
    slope end_minor;
    T cumulative_value;
    bool skip_first_row = false;
    bool skip_first_column = false;
};
```
[nitpick] doesn't seem to need constructor as it's plain struct